### PR TITLE
Speech chars v2

### PR DESCRIPTION
**Pull request by GeorgeEfstathiadis**
Speech Characteristics version 2 module

Methods doc: https://docs.google.com/document/d/1qmCbVMMUwOQP_hmNjXlZEHRCScwoF09-eZtlpT2RdM0/edit

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1204757386593242

### CODE REVIEW

**Comment by GeorgeEfstathiadis** [comment]
@vjbytes102 I think now it follows the guidelines from other functions more closely. When you merge the eye blink PR, I can use the general config file loading function here too.

File: openwillis/measures/text/speech_attribute.py [hunk 1]
```diff
@@ -4,154 +4,404 @@
...
-                                    measures['pos'], measures['compound'], measures['speech_mattr'], measures['rate_of_speech'],
-                                    measures['pause_rate'], measures['pause_meandur'], measures['silence_ratio']])
+    return "jobName" in json_conf and "results" in json_conf
 
-    # Create an empty tag dataframe
-    tag_df = pd.DataFrame(columns=[measures['word'], measures['tag']])
-    return tag_df, summ_df
 
-def is_amazon_transcribe(json_conf):
+def filter_speaker_label(item_data, speaker_label, turns_idxs, turns, phrases_idxs, phrases):
```
Comment (vjbytes102) [inline openwillis/measures/text/speech_attribute.py:158]: Modularization: will suggest to split this function into multiple small functions for clean code structure

File: openwillis/measures/text/speech_attribute.py [hunk 2]
```diff
@@ -4,154 +4,404 @@
...
+
+    return turns_idxs, turns, phrases_idxs, phrases
+
+
+def filter_transcribe(json_conf, measures, speaker_label=None):
     """
     ------------------------------------------------------------------------------------------------------
 
-    This function extracts the text and filters the JSON data for Amazon Transcribe json response objects.
+    This function extracts the text and filters the JSON data
```
Comment (vjbytes102) [inline openwillis/measures/text/speech_attribute.py:253]: same suggestion as above function:
- manage exception handling
- split it into smaller function
- try to include helper function in util file

File: openwillis/measures/text/speech_attribute.py [hunk 3]
```diff
@@ -4,154 +4,404 @@
...
 
-        # extract the text from the filtered json data
-        text_list = [item['alternatives'][0]['content'] for item in item_data if 'alternatives' in item]
-        text = " ".join(text_list)
+    # extract words
+    words = [word["alternatives"][0]["content"] for word in filter_json]
 
-    filter_json = [item for item in item_data if 'start_time' in item and 'end_time' in item]
-    return text, filter_json
+    return (
```
Comment (vjbytes102) [inline openwillis/measures/text/speech_attribute.py:365]: too many return values

Suggestion: save all of these in a list and return a single list object, and in speech_characteristics function use the list index to pass them as argument in process_language_feature

File: openwillis/measures/text/speech_attribute.py [hunk 4]
```diff
@@ -168,34 +418,76 @@ def speech_characteristics(json_conf, language='en-us', speaker_label=None):
...
+    measures = get_config(os.path.abspath(__file__), "text.json")
+    word_df, phrase_df, turn_df, summ_df = create_empty_dataframes(measures)
 
     try:
         if bool(json_conf):
             cutil.download_nltk_resources()
 
             if is_amazon_transcribe(json_conf):
-                text, filter_json = filter_transcribe(json_conf, speaker_label=speaker_label)
+                (
```
Comment (vjbytes102) [inline openwillis/measures/text/speech_attribute.py:440]: as stated above, it should be returned as single list object

File: openwillis/measures/text/speech_attribute.py [hunk 5]
```diff
@@ -168,34 +418,76 @@ def speech_characteristics(json_conf, language='en-us', speaker_label=None):
...
                 if len(text) > 0:
-                    tag_df, summ_df = cutil.process_language_feature(json_conf, [tag_df, summ_df], text, language, measures,
-                                                               ['start', 'end'])
-
+                    (
+                        word_df,
+                        phrase_df,
+                        turn_df,
+                        summ_df,
+                    ) = cutil.process_language_feature(
```
Comment (vjbytes102) [inline openwillis/measures/text/speech_attribute.py:473]: same as above

File: openwillis/measures/text/util/characteristics_util.py [hunk 6]
```diff
@@ -176,23 +236,209 @@ def get_mattr(word):
...
+    # column names
+    cols = [
+        measures["neg"],
+        measures["neu"],
+        measures["pos"],
+        measures["compound"],
+        measures["speech_mattr"],
+    ]
+
+    # word-level analysis
```
Comment (vjbytes102) [inline openwillis/measures/text/util/characteristics_util.py:279]: do we need to handle exception here?

File: openwillis/measures/text/speech_attribute.py [hunk 7]
```diff
@@ -4,154 +4,404 @@
...
 
     ------------------------------------------------------------------------------------------------------
     """
-    dir_name = os.path.dirname(os.path.abspath(__file__))
-    measure_path = os.path.abspath(os.path.join(dir_name, 'config/text.json'))
 
-    file = open(measure_path)
-    measures = json.load(file)
-    return measures
+    word_df = pd.DataFrame(
```
[the remark this answers is not in the document]
Comment (GeorgeEfstathiadis) [reply openwillis/measures/text/speech_attribute.py]: [move create_empty_dataframes to cutil](https://github.com/bklynhlth/openwillis/pull/22/commits/ad3ed23720af4eaf8f1148ee4534a61849ed4443)

File: openwillis/measures/text/speech_attribute.py (hunk 1 above)
Comment (GeorgeEfstathiadis) [reply openwillis/measures/text/speech_attribute.py]: [move filter_speaker to cutil and split into smaller functions](https://github.com/bklynhlth/openwillis/pull/22/commits/b0307bad85d4e86a1a7ee1064268e2d07fd88408)

File: openwillis/measures/text/speech_attribute.py (hunk 3 above)
Comment (GeorgeEfstathiadis) [reply openwillis/measures/text/speech_attribute.py]: [group output of filter_transcribe](https://github.com/bklynhlth/openwillis/pull/22/commits/10ee38b7b0caac580e10a1e0b072f6ae49795eea)

File: openwillis/measures/text/speech_attribute.py (hunk 4 above)
Comment (GeorgeEfstathiadis) [reply openwillis/measures/text/speech_attribute.py]: https://github.com/bklynhlth/openwillis/pull/22/commits/21195f7b6bf9c238b8fc06322e1d896fc34501ed

File: openwillis/measures/text/speech_attribute.py [hunk 8]
```diff
@@ -168,34 +418,76 @@ def speech_characteristics(json_conf, language='en-us', speaker_label=None):
...
-                    tag_df, summ_df = cutil.process_language_feature(filter_json, [tag_df, summ_df], text, language,
-                                                               measures, ['start_time', 'end_time'])
+                    (
+                        word_df,
+                        phrase_df,
+                        turn_df,
+                        summ_df,
+                    ) = cutil.process_language_feature(
+                        filter_json,
+                        [word_df, phrase_df, turn_df, summ_df],
```
[the remark this answers is not in the document]
Comment (GeorgeEfstathiadis) [reply openwillis/measures/text/speech_attribute.py]: https://github.com/bklynhlth/openwillis/pull/22/commits/21195f7b6bf9c238b8fc06322e1d896fc34501ed

File: openwillis/measures/text/speech_attribute.py (hunk 5 above)
Comment (GeorgeEfstathiadis) [reply openwillis/measures/text/speech_attribute.py]: https://github.com/bklynhlth/openwillis/pull/22/commits/21195f7b6bf9c238b8fc06322e1d896fc34501ed

File: openwillis/measures/text/util/characteristics_util.py (hunk 6 above)
Comment (GeorgeEfstathiadis) [reply openwillis/measures/text/util/characteristics_util.py:710]: [try-except block in sentiment analysis](https://github.com/bklynhlth/openwillis/pull/22/commits/ab90b9106ba9c0ba3bd91fc0d1e6bb9e81050a60)

File: openwillis/measures/text/util/characteristics_util.py [hunk 9]
```diff
@@ -176,23 +236,209 @@ def get_mattr(word):
...
+    """
+
+    if level_name not in [measures["phrase"], measures["turn"]]:
+        logger.error(
+            f"level_name must be either {measures['phrase']} or {measures['turn']}"
+        )
+        return df
+
+    for j, index in enumerate(index_list):
+        rng = range(index[0], index[1] + 1)
```
[the remark this answers is not in the document]
Comment (GeorgeEfstathiadis) [reply openwillis/measures/text/util/characteristics_util.py]: [try-except block in phrase/turn pause feature calculation](https://github.com/bklynhlth/openwillis/pull/22/commits/708189f1e37156e109c3764580438e7b009fa484)

File: openwillis/measures/text/speech_attribute.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (GeorgeEfstathiadis) [reply openwillis/measures/text/speech_attribute.py]: [group output of speech_characteristics to df_list](https://github.com/bklynhlth/openwillis/pull/22/commits/21195f7b6bf9c238b8fc06322e1d896fc34501ed)

**Review by vjbytes102 — APPROVED** [review]
Looks good

**Pull request merged**